src/openzaak/tests/utils/factories.py

```diff
@@ -0,0 +1,50 @@
+# SPDX-License-Identifier: EUPL-1.2
+# Copyright (C) 2022 Dimpact
+from furl import furl
+from zgw_consumers.constants import APITypes
+from zgw_consumers.models import Service
+
+from openzaak.utils.fields import FkOrServiceUrlField
+
+
+class FkOrServiceUrlFactoryMixin:
+    """
+    create service instances for composite url fields
+    create service for provided parameter 'local_host'
+    """
+
+    local_host = "http://testserver/"  # use to create service for local api
+
+    @classmethod
+    def _create(cls, model_class, *args, **kwargs):
+        """
+        post_generation hook can't be used since services should exist during factory.create()
+        """
+
+        services = []
+        fields = [
+            f.name
+            for f in model_class._meta.get_fields()
+            if isinstance(f, FkOrServiceUrlField)
+        ]
+        for field in fields:
+            value = kwargs.get(field)
+            if not value:
+                continue
+
+            if not isinstance(value, str):
+                continue
+
+            # create Service instance for composite field
+            base_url = furl(value).origin
+            services.append(Service(api_root=base_url, api_type=APITypes.orc))
+            # service, created = Service.objects.get_or_create(api_root=base_url, defaults={"api_type": APITypes.orc})
+
+        # if local host specified create service for it
+        local_host = kwargs.pop("local_host", cls.local_host)
+        if local_host:
+            services.append(Service(api_root=local_host, api_type=APITypes.orc))
```

Comment:
I really don't understand why a local service needs to exist before you can use Open Zaak? Surely this cannot be needed at runtime, as that would require people (now) to configure:

* `django.contrib.sites.Site` with the domain
* `settings.OPENZAAK_DOMAIN` with the domain
* add a `Service` for the domain that Open Zaak is hosted on

That's three times we're configuring the same thing :/

Reply:
Unfortunately `FkOrURLDescriptor` can't set string value without the related service. 
And afaik we use `FkOrURLDescriptor.__set__` with local url as string value in:
* `cmis_doc_to_django_model` function used for CMIS
* `virtual_model` initialization. For example, if we have ZIO, where informatieobject is external, but it has `informatieobjecttype`  which belongs to local catalogi. In this case we treat `informatieobjecttype` as external url

Adding `Service` data seemed the easiest (but of course not the prettiest) solution for me, otherwise it's the serios refactor of loose-fk descriptor, loader and/or virtual_model and I really didn't want to do it